x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/SecurityRestFilterTests.java

```diff
@@ -258,5 +277,100 @@ public void testSanitizeHeaders() throws Exception {
+                    }
+
+                    @Override
+                    public RestRequest checkRestPartial(RestHandler restHandler, RestRequest restRequest, ThreadContext threadContext) {
+                        return restRequest;
+                    }
+
+                    @Override
+                    public void maybeInterceptRequest(ThreadContext threadContext, TransportRequest request) {}
+                });
+
+                filter.handleRequest(request, channel, null);
+                if (isOperator) {
+                    verify(restHandler, times(1)).handleRequest(request, channel, null);
+                    verify(channel, never()).sendResponse(any()); // response is not sent from here
+                } else {
+                    verify(restHandler, never()).handleRequest(request, channel, null);
+                    verify(channel, times(1)).sendResponse(fullyRestrictedResponse); // response is sent from here
+                }
+            }
+        }
+    }
+
+    public void testPartiallyRestRestriction() throws Exception {
+        int i = 0;
+        for (Boolean isOperator : new Boolean[] { Boolean.TRUE, Boolean.FALSE }) {
+            i++;
+            RestRequest request = mock(RestRequest.class);
+            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                SecurityRestFilter filter = getFilter(new OperatorPrivileges.OperatorPrivilegesService() {
+                    @Override
+                    public void maybeMarkOperatorUser(Authentication authentication, ThreadContext threadContext) {}
+
+                    @Override
+                    public ElasticsearchSecurityException check(
+                        Authentication authentication,
+                        String action,
+                        TransportRequest request,
+                        ThreadContext threadContext
+                    ) {
+                        return null;
+                    }
+
+                    @Override
+                    public RestResponse checkRestFull(RestHandler restHandler, ThreadContext threadContext) {
+                        return null;
+                    }
+
+                    @Override
+                    public RestRequest checkRestPartial(RestHandler restHandler, RestRequest restRequest, ThreadContext threadContext) {
+                        return isOperator ? restRequest : mock(RestRequest.class);
+                    }
+
+                    @Override
+                    public void maybeInterceptRequest(ThreadContext threadContext, TransportRequest request) {}
+                });
+
+                filter.handleRequest(request, channel, null);
+                ArgumentCaptor<RestRequest> restRequestArgumentCaptor = ArgumentCaptor.forClass(RestRequest.class);
+                verify(restHandler, times(i)).handleRequest(restRequestArgumentCaptor.capture(), eq(channel), eq(null));
```

Comment:
Nit: I think we can get rid of the variable `i` if we call `Mockito.clearInvocations(restHandler)` before each loop.

Reply:
neat ! TIL (will update)